selfdrive/locationd/locationd.cc

```diff
@@ -0,0 +1,378 @@
+    VectorXd rot_device_std = (this->device_from_calib * floatlist2vector(log.getRotStd())) * 10.0;
+    this->kf->predict_and_observe(current_time, KIND_CAMERA_ODO_ROTATION,
+      { (VectorXd(rot_device.rows() + rot_device_std.rows()) << rot_device, rot_device_std).finished() });
+
+    VectorXd trans_device = this->device_from_calib * floatlist2vector(log.getTrans());
+    VectorXd trans_device_std = this->device_from_calib * floatlist2vector(log.getTransStd());
+
+    this->posenet_stds.pop_front();
+    this->posenet_stds.push_back(trans_device_std[0]);
+
+    trans_device_std *= 10.0;
+    this->kf->predict_and_observe(current_time, KIND_CAMERA_ODO_TRANSLATION,
+      { (VectorXd(trans_device.rows() + trans_device_std.rows()) << trans_device, trans_device_std).finished() });
+  }
+}
+
+void Localizer::handle_live_calib(double current_time, const cereal::LiveCalibrationData::Reader& log) {
+  if (log.getRpyCalib().size() > 0) {
+    this->calib = floatlist2vector(log.getRpyCalib());
+    this->device_from_calib = euler2rot(this->calib);
+    this->calib_from_device = this->device_from_calib.transpose();
+    this->calibrated = log.getCalStatus() == 1;
+  }
+}
+
+void Localizer::reset_kalman(double current_time) {
+  VectorXd init_x = this->kf->get_initial_x();
+  this->reset_kalman(current_time, init_x.segment<4>(3), init_x.head(3));
+}
+
+void Localizer::reset_kalman(double current_time, VectorXd init_orient, VectorXd init_pos) {
+  // too nonlinear to init on completely wrong
+  VectorXd init_x = this->kf->get_initial_x();
+  MatrixXdr init_P = this->kf->get_initial_P();
+  init_x.segment<4>(3) = init_orient;
+  init_x.head(3) = init_pos;
+
+  this->kf->init_state(init_x, init_P, current_time);
+
+  this->gyro_counter = 0;
+  this->acc_counter = 0;
+  this->speed_counter = 0;
+  this->cam_counter = 0;
+}
+
+void Localizer::handle_msg_bytes(const char *data, const size_t size) {
+  AlignedBuffer aligned_buf;
+
+  capnp::FlatArrayMessageReader cmsg(aligned_buf.align(data, size));
+  cereal::Event::Reader event = cmsg.getRoot<cereal::Event>();
+
+  this->handle_msg(event);
+}
+
+void Localizer::handle_msg(const cereal::Event::Reader& log) {
+  double t = log.getLogMonoTime() * 1e-9;
+  if (log.isSensorEvents()) {
+    this->handle_sensors(t, log.getSensorEvents());
+  } else if (log.isGpsLocationExternal()) {
+    this->handle_gps(t, log.getGpsLocationExternal());
+  } else if (log.isCarState()) {
+    this->handle_car_state(t, log.getCarState());
+  } else if (log.isCameraOdometry()) {
+    this->handle_cam_odo(t, log.getCameraOdometry());
+  } else if (log.isLiveCalibration()) {
+    this->handle_live_calib(t, log.getLiveCalibration());
+  }
+}
+
+kj::ArrayPtr<capnp::byte> Localizer::getMessageBytes(MessageBuilder& msg_builder, uint64_t logMonoTime,
+  bool inputsOK, bool sensorsOK, bool gpsOK)
+{
+  cereal::Event::Builder evt = msg_builder.initEvent();
+  evt.setLogMonoTime(logMonoTime);
```

Comment:
need to keep for this PR to make sure refs are the same, but any reason for this? if we want this info, we should probably just add a field to log the cameraOdometry time separately

Reply:
I think this might have been done to make use in the pipeline easier. We should probably just change the timestamps in the stage itself.